src/lib/schedule.c

```diff
@@ -25,459 +25,95 @@
 {
-	struct schedule_data *sch = *arch_schedule_get();
-	uint32_t flags;
 	int ret = 0;
 
-	tracev_schedule("schedule_task_cancel()");
-
-	spin_lock_irq(&sch->lock, flags);
-
-	/* check current task state, delete it if it is queued
-	 * if it is already running, nothing we can do about it atm
-	 */
-	if (task->state == TASK_STATE_QUEUED) {
-		/* delete task */
-		task->state = TASK_STATE_CANCEL;
-		list_item_del(&task->list);
-	}
-
-	spin_unlock_irq(&sch->lock, flags);
+	if (schedulers[task->type]->schedule_task_cancel)
+		ret = schedulers[task->type]->schedule_task_cancel(task);
 
 	return ret;
 }
 
-static int _schedule_task(struct task *task, uint64_t start, uint64_t deadline)
-{
-	struct schedule_data *sch = *arch_schedule_get();
-	uint32_t flags;
-	uint64_t current;
-	uint64_t ticks_per_ms;
-
-	tracev_schedule("_schedule_task()");
-
-	spin_lock_irq(&sch->lock, flags);
-
-	/* is task already pending ? - not enough MIPS to complete ? */
-	if (task->state == TASK_STATE_PENDING) {
-		trace_schedule("_schedule_task(), task already pending");
-		spin_unlock_irq(&sch->lock, flags);
-		return 0;
-	}
-
-	/* is task already queued ? - not enough MIPS to complete ? */
-	if (task->state == TASK_STATE_QUEUED) {
-		trace_schedule("_schedule_task(), task already queued");
-		spin_unlock_irq(&sch->lock, flags);
-		return 0;
-	}
-
-	/* get the current time */
-	current = platform_timer_get(platform_timer);
-
-	ticks_per_ms = clock_ms_to_ticks(sch->clock, 1);
-
-	/* calculate start time - TODO: include MIPS */
-	if (start == 0)
-		task->start = current;
-	else
-		task->start = task->start + ticks_per_ms * start / 1000 -
-			PLATFORM_SCHEDULE_COST;
-
-	/* calculate deadline - TODO: include MIPS */
-	task->deadline = task->start + ticks_per_ms * deadline / 1000;
-
-	/* add task to list */
-	list_item_append(&task->list, &sch->list);
-	task->state = TASK_STATE_QUEUED;
-	spin_unlock_irq(&sch->lock, flags);
-
-	return 1;
-}
-
-/*
- * Add a new task to the scheduler to be run and define a scheduling
- * deadline in time for the task to be ran. Do not invoke the scheduler
- * immediately to run task, but wait intil schedule is next called.
- *
- * deadline is in microseconds relative to start.
- */
-void schedule_task_idle(struct task *task, uint64_t deadline)
-{
-	_schedule_task(task, 0, deadline);
-}
-
-/*
- * Add a new task to the scheduler to be run and define a scheduling
- * window in time for the task to be ran. i.e. task will run between start and
- * deadline times.
- *
- * start is in microseconds relative to last task start time.
- * deadline is in microseconds relative to start.
- */
-void schedule_task(struct task *task, uint64_t start, uint64_t deadline)
+int scheduler_init(void)
 {
-	int need_sched;
-
-	need_sched = _schedule_task(task, start, deadline);
+	int i = 0;
+	int ret = 0;
 
-	/* need to run scheduler if task not already running */
-	if (need_sched) {
-		/* rerun scheduler */
-		schedule();
+	for (i = 0; i < SOF_TASKS_COUNT; i++) {
```

Comment:
SOF_TASKS_COUNT is the number of schedulers ? naming makes no sense.

Reply:
Changed to ```SOF_SCHEDULE_COUNT```